File: solutions/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        current = head
        delayed = head
        i = 0
        while current:
            current = current.next
            if i > n:
                delayed = delayed.next
            i += 1

        if i == n:
            return head.next

        if delayed.next:
            delayed.next = delayed.next.next

        return head
        
# @lc code=end

File: solutions/19.remove-nth-node-from-end-of-list.java
/*
 * @lc app=leetcode id=19 lang=java
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode current = head;
        ListNode delay = head;
        int i = 0;
        while (current != null) {
            current = current.next;
            if (i++ > n) {
                delay = delay.next;
            }
        }
        if (i == n) {
            return head.next;
        }
        if (delay.next != null) {
            delay.next = delay.next.next;
        }
        return head;
    }
}
// @lc code=end

